presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PushdownFilterIntoWindow.java

```diff
@@ -0,0 +1,184 @@
+import static io.prestosql.SystemSessionProperties.isOptimizeTopNRowNumber;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.spi.predicate.Marker.Bound.BELOW;
+import static io.prestosql.spi.predicate.Range.range;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.sql.planner.DomainTranslator.fromPredicate;
+import static io.prestosql.sql.planner.plan.Patterns.filter;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.window;
+import static java.lang.Math.toIntExact;
+
+public class PushdownFilterIntoWindow
+        implements Rule<FilterNode>
+{
+    private final Capture<WindowNode> childCapture;
+    private final Pattern<FilterNode> pattern;
+
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+    private final FunctionId rowNumberFunctionId;
+
+    public PushdownFilterIntoWindow(Metadata metadata)
+    {
+        this.metadata = metadata;
+        this.domainTranslator = new DomainTranslator(metadata);
+        this.rowNumberFunctionId = metadata.resolveFunction(QualifiedName.of("row_number"), ImmutableList.of()).getFunctionId();
+        this.childCapture = newCapture();
+        this.pattern = filter()
+                .with(source().matching(window()
+                        .matching(window -> window.getWindowFunctions().size() == 1)
+                        .matching(window -> getOnlyElement(window.getWindowFunctions().values()).getResolvedFunction().getFunctionId().equals(rowNumberFunctionId))
+                        .capturedAs(childCapture)));
+    }
+
+    @Override
+    public Pattern<FilterNode> getPattern()
+    {
+        return pattern;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isOptimizeTopNRowNumber(session);
+    }
+
+    @Override
+    public Result apply(FilterNode node, Captures captures, Context context)
+    {
+        Session session = context.getSession();
+        TypeProvider types = context.getSymbolAllocator().getTypes();
+        PlanNodeIdAllocator idAllocator = context.getIdAllocator();
+
+        WindowNode windowNode = captures.get(childCapture);
+
+        TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), types).getTupleDomain();
+
+        Symbol rowNumberSymbol = getOnlyElement(windowNode.getWindowFunctions().entrySet()).getKey();
+        OptionalInt upperBound = extractUpperBound(tupleDomain, rowNumberSymbol);
+
+        if (upperBound.isPresent()) {
+            if (upperBound.getAsInt() <= 0) {
+                return Result.ofPlanNode(new ValuesNode(node.getId(), node.getOutputSymbols(), ImmutableList.of()));
+            }
+            if (windowNode.getOrderingScheme().isPresent()) {
+                TopNRowNumberNode newSource = new TopNRowNumberNode(idAllocator.getNextId(),
```

Comment:
Please reuse the `id` of the `WindowNode`, and move the argument to a new line.
Then you don't need `idAllocator`.